XRSessionMode proposal

toji opened the pull request:
Finding this one tough to keep focused on JUST introducing the `XRSessionMode` concept, but that's probably necessary for not getting lost in the weeds, Gonna add some comments inline re: thing I'm thinking about that I'd like opinions on.

toji commented on explainer.md:
```diff
@@ -95,17 +95,19 @@ navigator.xr.addEventListener('devicechange', checkForXR);
+A session can be created with one of three modes:
 
-**Immersive**: Requested with the `immersive: true` dictionary argument. Immersive sessions present content directly to the `XRDevice`, enabling immersive presentation. Only one immersive session per XR hardware device is allowed at a time across the entire UA. Immersive sessions must be created within a user activation event or within another callback that has been explicitly indicated to allow  ...
+**Inline**: The default mode, but can be explicitly requested with the `mode: 'inline'` dictionary argument. Inline sessions do not have the ability to display immersive content on the `XRDevice` but are able to access device tracking information and use it to render content on the page. This technique, where a scene rendered to the page is responsive to device movement, is sometimes referred to  ...
```
There's a few things I want to do here, eventually. One is scrub this document of the term "magic window" and focus entirely on advancing understanding of the "inline" concept. The other is restructure the entire doc so that inline and AR mode aren't presented as footnotes in the "advanced" section. I think that shift may be more appropriate when we tackle the Universal Rendering Path concept, though.

toji commented on explainer.md:
```diff
@@ -95,17 +95,19 @@ navigator.xr.addEventListener('devicechange', checkForXR);
+**Inline**: The default mode, but can be explicitly requested with the `mode: 'inline'` dictionary argument. Inline sessions do not have the ability to display immersive content on the `XRDevice` but are able to access device tracking information and use it to render content on the page. This technique, where a scene rendered to the page is responsive to device movement, is sometimes referred to  ...
 
-**Non-Immersive (in-page)**: The default mode, but can be explicitly requested with the `immersive: false` dictionary argument. Non-immersive sessions do not have the ability to display immersive content on the `XRDevice` but are able to access device tracking information and use it to render content on the page. This technique, where a scene rendered to the page is responsive to device movement, ...
+**Immersive**: Requested with the `mode: 'immersive'` dictionary argument. Immersive sessions present content directly to the XR hardware device (frequently a headset), enabling immersive presentation. Immersive sessions are considered "exclusive". Only one exclusive session per XR hardware device is allowed at a time across the entire UA, and any non-exclusive sessions are suspended while an exc ...
```
Sticking with the term "immersive" for the moment to avoid conflating this with the ongoing bikeshed, but I'm not particularly attached to the term. It is a bit weird to have to explain that the immersive modes are exclusive in how they access the hardware, so perhaps reverting to that term simplifies the message.

toji replied:
> Actually, it totally doesn't have to be and that's key. There's absolutely no reason an AR Core device would need to be in Daydream mode. Just sayin'

I hear what you're saying, but I can't help but wonder about the developer intent behind it. If I put a button on your page that looks like a headset, but when you click it all you get is a fancy fullscreen mode with non-obvious 6DoF tracking it's likely not what you or the developer wanted? (It's frequently hard to differentiate between orientation tracking and true 6DoF when you're sitting in place and viewing a life-sized scene through a handheld screen.)

Plus, if that's the behavior you wanted it seems like you could just ask for immersive-ar and draw over the full screen? (Handwaving about lost performance.)

toji commented on explainer.md:
```diff
@@ -95,17 +95,19 @@ navigator.xr.addEventListener('devicechange', checkForXR);
-**Non-Immersive (in-page)**: The default mode, but can be explicitly requested with the `immersive: false` dictionary argument. Non-immersive sessions do not have the ability to display immersive content on the `XRDevice` but are able to access device tracking information and use it to render content on the page. This technique, where a scene rendered to the page is responsive to device movement, ...
+**Immersive**: Requested with the `mode: 'immersive'` dictionary argument. Immersive sessions present content directly to the XR hardware device (frequently a headset), enabling immersive presentation. Immersive sessions are considered "exclusive". Only one exclusive session per XR hardware device is allowed at a time across the entire UA, and any non-exclusive sessions are suspended while an exc ...
+
+**Immersive AR**: Requested with the `mode: 'immersive-ar'` dictionary argument. These sessions function largely like standard Immersive sessions, with the primary difference being that they guarantee that the users environment will be visible and aligned with the rendered content. (It should also be noted that standard immersive sessions may make the environment visible anyway, especially on tra ...
```
Would *love* a sanity check on if this description of AR mode (and the additional details below) actually click. I feel like I'm being overly simplistic.

Also when we start talking about "Immersive AR" I find myself backed into talking about "standard immersive sessions", which.... ewe.

toji commented on explainer.md:
```diff
@@ -95,17 +95,19 @@ navigator.xr.addEventListener('devicechange', checkForXR);
 If an `XRDevice` is available and able to create an immersive session, the application will usually want to add some UI to trigger activation of "XR Presentation Mode", where the application can begin sending imagery to the device. Testing to see if the device supports the capabilities the application needs is done via the `supportsSession` call, which takes a dictionary of the desired functional ...
 
-In the following examples we will focus on using immersive sessions, and cover non-immersive session use in the [`Advanced Functionality`](#non-immersive-sessions-magic-windows) section. With that in mind, we ask here if the `XRDevice` supports immersive sessions, since we want the ability to display imagery on the headset.
+In the following examples we will focus on using immersive sessions, and cover inline session use in the [`Advanced Functionality`](#inline-sessions-magic-windows) section. With that in mind, we ask here if the `XRDevice` supports immersive sessions, since we want the ability to display imagery on the headset.
```
LOTS of straightforward "non-immersive"->"inline" search and replace here. Apologies for the noise.

toji commented on explainer.md:
````diff
@@ -346,50 +348,69 @@ function beginXRSession() {
+}
+```
+
+In addition to displaying on a dedicated XR hardware device like a standard Immersive session, the UA may choose instead to display an AR session output on a 2D screen, such as with [ARKit](https://developer.apple.com/arkit/) and [ARCore](https://developers.google.com/ar/) compatible devices. When displayed in that way, the UA must transition to a mode where the AR session's output is shown exclu ...
````
A bunch of questions this paragraph brings up for me.

Do we need to communicate to the developer that they're going to be presenting to a 2D screen rather than a headset? If so should that be before or after session creation? Is that pit-of-success-y? Until we add DOM overlays (which I'm studiously avoiding talking about here) it largely feels like the main difference between headset AR and phone AR from a developer's perspective is the available input devices.

And FWIW when it comes to DOM overlays and such I still think there's possibly room for a 4th mode that's explicitly targeted at Phone AR-style interaction, but if we went that route I'd say that it would fit comfortably as a 4th mode in the enum ('immersive-ar-with-overlay' or something silly like that) and that this mode could still function exactly as described.

NellWaliczek replied:
I'd leave it out any references to the overlay or differentiating handheld devices for now.  Let's keep this change scoped, but make sure we've got an issue filed to cover it.

toji commented on explainer.md:
```diff
@@ -1008,14 +1029,20 @@ partial interface Navigator {
 [SecureContext, Exposed=Window] interface XRSession : EventTarget {
   readonly attribute XRDevice device;
-  readonly attribute boolean immersive;
+  readonly attribute XRSessionMode mode;
```
We both agreed previously that "mode" makes the most sense when the sessions are mutable, which is still something I plan to propose but am avoiding here. Sticking with the "mode" terminology anyway to potentially reduce churn.

NellWaliczek replied:
I wonder if that'll end up causing a bikeshedding rathole... no strong opinion right now

toji replied:
Even if we're avoiding the mutable sessions path for the time being now, I've come around to "mode" being the more natural thing to type in many cases, so I'm sticking to it.

NellWaliczek commented on explainer.md:
````diff
@@ -346,50 +348,69 @@ function beginXRSession() {
 ```
 
-The UA may reject requests for a non-immersive sessions for a variety of reasons, such as the inability of the underlying hardware to provide tracking data without actively rendering to the device. Pages should be designed to robustly handle the inability to acquire non-immersive sessions. `XRDevice.supportsSession()` can be used if a page wants to test for non-immersive session support before at ...
+The UA may reject requests for an inline sessions for a variety of reasons, such as the inability of the underlying hardware to provide tracking data without actively rendering to the device. Pages should be designed to robustly handle the inability to acquire inline sessions. `XRDevice.supportsSession()` can be used if a page wants to test for inline session support before attempting to create t ...
````
Do we want to require inline sessions to be supported?  Or do we need to resolve the poseless sessions thing first?

toji replied:
I wondered about that. Felt to me like we needed to resolve poseless first.

NellWaliczek commented on explainer.md:
```diff
@@ -432,7 +453,7 @@ The `targetRay` will never be `null`. It's value will differ based on the type o
   * `'gaze'` indicates the target ray will originate at the user's head and follow the direction they are looking (this is commonly referred to as a "gaze input" device). While it may be possible for these devices to be tracked (and have a grip matrix), the head gaze is used for targeting. Example devices: 0DOF clicker, regular gamepad, voice command, tracked hands.
   * `'tracked-pointer'` indicates that the target ray originates from either a handheld device or other hand-tracking mechanism and represents that the user is using their hands or the held device for pointing. The exact orientation of the ray relative to a given device should follow platform-specific guidelines if there are any. In the absence of platform-specific guidance or a physical device,  ...
-  * `'screen'` indicates that the input source was an interaction with the canvas element associated with a non-immersive session's output context, such as a mouse click or touch event. See [Magic Window Input](#magic_window_input) for more details.
+  * `'screen'` indicates that the input source was an interaction with the canvas element associated with a inline session's output context or an immersive AR session, such as a mouse click or touch event. See [Magic Window Input](#magic_window_input) for more details.
```
Technically, you could build a VR session on ARCore....

toji commented on explainer.md:
```diff
@@ -95,17 +95,19 @@ navigator.xr.addEventListener('devicechange', checkForXR);
+A session can be created with one of three modes:
 
-**Immersive**: Requested with the `immersive: true` dictionary argument. Immersive sessions present content directly to the `XRDevice`, enabling immersive presentation. Only one immersive session per XR hardware device is allowed at a time across the entire UA. Immersive sessions must be created within a user activation event or within another callback that has been explicitly indicated to allow  ...
+**Inline**: The default mode, but can be explicitly requested with the `mode: 'inline'` dictionary argument. Inline sessions do not have the ability to display immersive content on the `XRDevice` but are able to access device tracking information and use it to render content on the page. This technique, where a scene rendered to the page is responsive to device movement, is sometimes referred to  ...
```
Yeah, that's one of those "definitely intend to change this but trying to keep this PR focused" things. (I'm finding a LOT of those as I go)

toji commented:
Rebased and updated. I **know** I'm gonna get some people up in arms with the `immersive-vr` and `immersive-ar` enum suggestions, but honestly the more I think about it the more I think we really just need names that suggest to the developer what they WANT to do, rather than how the system sees it. (This mentality inspired by naming Frames of Reference things like 'stationary', because that's how you expect the user to experience your content.)

Would appreciate another look, and if there's nothing terrifying I'll push the PR to the `webxr` repo to watch people tear each other apart over the mode names. (Don't intend to merge prior to the FoR changes landing if I can help it.)

NellWaliczek commented on explainer.md:
```diff
@@ -410,7 +431,7 @@ The `targetRay` will never be `null`. It's value will differ based on the type o
   * `'gaze'` indicates the target ray will originate at the user's head and follow the direction they are looking (this is commonly referred to as a "gaze input" device). While it may be possible for these devices to be tracked (and have a grip matrix), the head gaze is used for targeting. Example devices: 0DOF clicker, regular gamepad, voice command, tracked hands.
   * `'tracked-pointer'` indicates that the target ray originates from either a handheld device or other hand-tracking mechanism and represents that the user is using their hands or the held device for pointing. The exact orientation of the ray relative to a given device should follow platform-specific guidelines if there are any. In the absence of platform-specific guidance or a physical device,  ...
-  * `'screen'` indicates that the input source was an interaction with the canvas element associated with a non-immersive session's output context, such as a mouse click or touch event. See [Magic Window Input](#magic_window_input) for more details.
+  * `'screen'` indicates that the input source was an interaction with the canvas element associated with a inline session's output context or an immersive AR session, such as a mouse click or touch event. See [Inline Input](#inline_input) for more details.
```
This is only true if the immersive AR session isn't in an headset, yeah?

NellWaliczek replied:
A general comment... do we need to provide some indication of the session not being headworn for UI purposes?  Or is that being punted to the DOM overlay issue?

toji replied:
That specific spec issue feels like it's better handled in the context of the DOM overlay issue, yes.

toji commented on explainer.md:
```diff
@@ -75,21 +75,23 @@ The first thing that any XR-enabled page will want to do is query to determine i
-In the following examples we will focus on using immersive content, and cover non-immersive content use in the [`Advanced Functionality`](#non-immersive-sessions-magic-windows) section. With that in mind, this code checks for supports of immersive content, since we want the ability to display imagery on a device like a headset.
+**Inline**: The default mode, but can be explicitly requested with the `mode: 'inline'` dictionary argument. inline content does not have the ability to display on the XR device, but is able to access device tracking information and use it to render content on the page. This technique, where a scene rendered to the page is responsive to device movement, is sometimes referred to as "Magic Window"  ...
+
+In the following examples we will focus on using immersive VR content, and cover inline and immersive AR content use in the [`Advanced Functionality`](#inline-sessions) section. With that in mind, this code checks for supports of immersive VR content, since we want the ability to display imagery on a device like a headset.
```
That was my thoughts on it. (Re: waiting for mutable output contexts) General concern is that I want to introduce the simplest mode with the least amount of setup first, which happens to be immersive VR. That way the document doesn't start off with the already confusing concept of output contexts. When we have mutable output contexts (and especially if we make WebGL contexts a valid output context) that progression will hopefully become less problematic.

toji commented on explainer.md:
````diff
@@ -120,7 +122,7 @@ function beginXRSession() {
 ```
 
-Only one immersive session per XR hardware device is allowed at a time across the entire UA. Any non-immersive sessions are suspended when an immersive session is active. Non-immersive sessions are not required to be created within a user activation event unless paired with another option that explicitly does require it.
+Only one immersive session per XR hardware device is allowed at a time across the entire UA. Any inline sessions are suspended when an immersive session is active. Inline sessions are not required to be created within a user activation event unless paired with another option that explicitly does require it.
````
Added a note early in the document to explain this.

toji commented:
Updated again, being a bit more aggressive about reworking how certain concepts are presented now while still steering clear of a full document restructuring. I'm hoping to defer moving the inline sections around until we have the mutable outputContext changes in place.

NellWaliczek commented on explainer.md:
```diff
@@ -100,17 +108,17 @@ async function checkForXRSupport() {
+After confirming that the desired mode is available with `navigator.xr.supportsSessionMode()`, the application will need to request an `XRSession` instance in order to interact with XR device's presentation or tracking capabilities.
 
-Clicking the button in the previous sample will attempt to acquire an `XRSession` by calling `navigator.xr.requestSession()` method. This returns a promise that resolves to an `XRSession` upon success. When requesting a session, the capabilities that the returned session must have are passed in via a dictionary, exactly like the `supportsSession` call. If `supportsSession` resolved for a given di ...
+Clicking the button in the previous sample will attempt to acquire an `XRSession` by calling `navigator.xr.requestSession()` method. This returns a promise that resolves to an `XRSession` upon success. When requesting a session, the capabilities that the returned session must have, including it's XR mode, are passed in via a dictionary.
```
nit: this block and the paragraph above seem to be a bit duplicative.

NellWaliczek commented on explainer.md:
```diff
@@ -120,7 +128,9 @@ function beginXRSession() {
-Only one immersive session per XR hardware device is allowed at a time across the entire UA. Any non-immersive sessions are suspended when an immersive session is active. Non-immersive sessions are not required to be created within a user activation event unless paired with another option that explicitly does require it.
+If `supportsSessionMode` resolved for a given mode, then requesting a session with the same mode should be reasonably expected to succeed, barring external factors (such as `requestSession` not being called in a user activation event for an immersive session.) The UA is ultimately responsible for determining if it can honor the request.
+
+Only one immersive session per XR hardware device is allowed at a time across the entire UA. Any inline sessions are suspended when an immersive session is active. Inline sessions are not required to be created within a user activation event unless paired with another option that explicitly does require it.
```
nit: Any inline -> All inline

NellWaliczek commented on explainer.md:
```diff
@@ -246,7 +256,7 @@ function drawScene(view, pose) {
 The UA may temporarily "suspend" a session at any time. While suspended a session has restricted or throttled access to the XR device state and may process frames slowly or not at all. Suspended sessions can be reasonably be expected to be resumed at some point, usually when the user has finished performing whatever action triggered the suspension in the first place.
 
-The UA may suspend a session if allowing the page to continue reading the headset position represents a security or privacy risk (like when the user is entering a password or URL with a virtual keyboard, in which case the head motion may infer the user's input), or if other content is obscuring the page's output. Additionally, non-immersive sessions are suspended while an immersive session is act ...
+The UA may suspend a session if allowing the page to continue reading the headset position represents a security or privacy risk (like when the user is entering a password or URL with a virtual keyboard, in which case the head motion may infer the user's input), or if other content is obscuring the page's output. Additionally, inline sessions are suspended while an immersive session is active.
```
>Additionally, inline sessions are suspended while an immersive session is active.

A bit weird that we're saying this in two places, but ok :)

NellWaliczek commented on explainer.md:
````diff
@@ -294,17 +304,45 @@ function onSessionEnd() {
 ```
 
-The UA may end a session at any time for a variety of reasons. For example: The user may forcibly end presentation via a gesture to the UA, other native applications may take exclusive access of the XR hardware device, or the XR hardware device may become disconnected from the system. Additionally, if the system's underlying XR device changes (signaled by the `devicechange` event on the `navigato ...
+The UA may end a session at any time for a variety of reasons. For example: The user may forcibly end presentation via a gesture to the UA, other native applications may take exclusive access of the XR hardware device, or the XR hardware device may become disconnected from the system. Additionally, if the system's underlying XR device changes (signaled by the `devicechange` event on the `navigato ...
````
>This applies to both Immersive and Inline sessions.

I wish I remember why we decided this... It seems like this might be a pit of failure....  as in, developers won't monitor the event and the page will be very badly behaved if a headset is plugged in after the fact.

NellWaliczek commented on explainer.md:
````diff
@@ -294,17 +304,45 @@ function onSessionEnd() {
+}
+```
+
+This provides a session that behaves much like the typical Immersive VR sessions described above with a few key behavioral differences. The primary distinction between an "immersive-vr" and "immersive-ar" session is that the latter guarantees that the user's environment is visible and that rendered content will be aligned to the environment. The exact nature of the visibility is hardware-dependen ...
````
nit: the typical Immersive VR -> the Immersive VR

NellWaliczek commented on explainer.md:
````diff
@@ -294,17 +304,45 @@ function onSessionEnd() {
+}
+```
+
+The UA may choose to present the immersive AR session's content via any type of display, including dedicated XR hardware (for devices like HoloLens or Magic Leap) or 2D screens (for APIs like [ARKit](https://developer.apple.com/arkit/) and [ARCore](https://developers.google.com/ar/)). In all cases the session takes exclusive control of the display, hiding the rest of the page if necessary. On a p ...
````
I really like how  you reworded this paragraph 👍

NellWaliczek commented on explainer.md:
```diff
@@ -324,59 +362,59 @@ function beginXRSession() {
+Similar to mirroring, to make use of this mode an `XRPresentationContext` is provided as the `outputContext` at session creation time with an inline session. At that point content rendered to the `XRSession`'s `baseLayer` will be rendered to the canvas associated with the `outputContext`. The UA is also allowed to composite in additional content if desired. In the future, if multiple `XRLayers` a ...
 
-Immersive and non-immersive sessions can use the same render loop, but there are some differences in behavior to be aware of. The sessions may run their render loops at at different rates. During immersive sessions the UA runs the rendering loop at the XR device's native refresh rate. During non-immersive sessions the UA runs the rendering loop at the refresh rate of page (aligned with `window.re ...
+Immersive and inline sessions can use the same render loop, but there are some differences in behavior to be aware of. The sessions may run their render loops at at different rates. During immersive sessions the UA runs the rendering loop at the XR device's native refresh rate. During inline sessions the UA runs the rendering loop at the refresh rate of page (aligned with `window.requestAnimation ...
```
Oh geez.. this just made me realize another issue with the Diorama frame of reference... the refresh rate may be wrong O_O

NellWaliczek commented on explainer.md:
````diff
@@ -324,59 +362,59 @@ function beginXRSession() {
 ```
 
-The UA may reject requests for a non-immersive sessions for a variety of reasons, such as the inability of the underlying hardware to provide tracking data without actively rendering to the device. Pages should be designed to robustly handle the inability to acquire non-immersive sessions. `navigator.xr.supportsSession()` can be used if a page wants to test for non-immersive session support befor ...
+The UA should not reject requests for an inline session unless the page's feature policy prevents it. `navigator.xr.supportsSessionMode()` can still be used if a page wants to test if inline session are allowed.
````
Interesting... this seems like a new change... can you give me a bit of background on this change?

toji replied:
Are you referring specifically to the feature policy bit? I felt that was fairly natural conclusion given the way feature policies work. It would be weird, to me, if the parent page of an iframe said "No XR for you!" and the page was still happily able to create `XRSessions`, regardless of functionality level.

NellWaliczek commented on explainer.md:
```diff
@@ -410,7 +448,7 @@ The `targetRay` will never be `null`. It's value will differ based on the type o
   * `'gaze'` indicates the target ray will originate at the user's head and follow the direction they are looking (this is commonly referred to as a "gaze input" device). While it may be possible for these devices to be tracked (and have a grip matrix), the head gaze is used for targeting. Example devices: 0DOF clicker, regular gamepad, voice command, tracked hands.
   * `'tracked-pointer'` indicates that the target ray originates from either a handheld device or other hand-tracking mechanism and represents that the user is using their hands or the held device for pointing. The exact orientation of the ray relative to a given device should follow platform-specific guidelines if there are any. In the absence of platform-specific guidance or a physical device,  ...
-  * `'screen'` indicates that the input source was an interaction with the canvas element associated with a non-immersive session's output context, such as a mouse click or touch event. See [Magic Window Input](#magic_window_input) for more details.
+  * `'screen'` indicates that the input source was an interaction with a session's output context canvas element, such as a mouse click or touch event. Only applicable for inline sessions or an immersive AR session being displayed on a 2D screen. See [Screen Input](#screen_input) for more details.
```
Maybe add "instead of from within a headset" to make it really clear?

toji replied:
Augh! But what if the "2D screen" is a floating quad that you're viewing in your headset?!? WHY ARE THESE THINGS SO HARD TO DESCRIBE ACCURATELY?????

NellWaliczek commented on explainer.md:
```diff
@@ -649,9 +687,9 @@ function onUpdateScene() {
+### Screen Input
 
-When using a non-immersive session, pointer events on the canvas that created the `outputContext` passed during the session request are monitored. `XRInputSource`s are generated in response to allow unified input handling with immersive mode controller or gaze input.
+When using an inline session or an immersive AR session on a 2D screen, pointer events on the canvas that created the session's `outputContext` are monitored. `XRInputSource`s are generated in response to allow unified input handling with immersive mode controller or gaze input.
```
same as above

toji commented:
Moved to main repo and merged. Closing.
